main/src/library/Iterator.flix

```diff
@@ -14,179 +14,158 @@
-              if (j < n) {
-                i := j + 1;
-                x
-              }
-              else {
-                bug!("Empty Iterator!")
-              }
-            };
-        Iterator(done, next)
+    pub def repeat(rc: Region[r], n: Int32, x: a): Iterator[a, ef, r] \ Write(r) =
+        let ix = ref n @ rc;
+        let f = () -> {
+            let i = deref ix;
+            if (i < 1)
+                Done
+            else {
+                ix := i-1;
+                Ans(x)
+            }
+        };
+        let iterF = () -> unsafe_cast f() as _ \ { ef, r };
+        Iterator(rc, iterF)
 
     ///
     /// Returns the sum of all elements in the iterator `iter`.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def sum(iter: Iterator[Int32, r]): Int32 \ Read(r) =
+    pub def sum(iter: Iterator[Int32, ef, r]): Int32 \ { ef, Read(r) } =
         foldLeft((acc, x) -> acc + x, 0, iter)
 
     ///
     /// Returns the sum of all elements in the iterator `iter` according to the function `f`.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def sumWith(f: a -> Int32 \ ef, iter: Iterator[a, r]): Int32 \ { ef, Read(r) } =
+    pub def sumWith(f: a -> Int32 \ ef1, iter: Iterator[a, ef, r]): Int32 \ { ef, ef1, Read(r) } =
         foldLeft((acc, x) -> acc + f(x), 0, iter)
 
     ///
     /// Returns the product of all elements in the iterator `iter`.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def product(iter: Iterator[Int32, r]): Int32 \ Read(r) =
-        if (isEmpty(iter))
-            1
-        else
-            foldLeft((acc, x) -> acc * x, 1, iter)
+    pub def product(iter: Iterator[Int32, ef, r]): Int32 \ { ef, Read(r) } = match next(iter) {
+        case Some(i) => foldLeft((acc, x) -> acc * x, i, iter)
+        case None    => 1
+    }
 
     ///
     /// Returns the product of all elements in the iterator `iter` according to the function `f`.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def productWith(f: a -> Int32 \ ef, iter: Iterator[a, r]): Int32 \ { ef, Read(r) } =
-        if (isEmpty(iter))
-            1
-        else
-            foldLeft((acc, x) -> acc * f(x), 1, iter)
+    pub def productWith(f: a -> Int32 \ ef1, iter: Iterator[a, ef, r]): Int32 \ { ef, ef1, Read(r) } = match next(iter) {
+        case Some(a) => foldLeft((acc, x) -> acc * f(x), f(a), iter)
+        case None    => 1
+    }
 
     ///
     /// Returns the contents of `iter` as an array.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def toArray(r1: Region[r1], iter: Iterator[a, r2]): Array[a, r1] \ { Read(r2), Write(r1) } = region r0 {
-        let m = new MutList(r0);
+    pub def toArray(rc: Region[r1], iter: Iterator[a, ef, r2]): Array[a, r1] \ { ef, Read(r2), Write(r1) } = region rc1 {
+        let m = new MutList(rc1);
         forEach(a -> MutList.push!(a, m), iter);
-        MutList.toArray(r1, m)
+        MutList.toArray(rc, m)
```

Comment:
Very nice, should in theory be very efficient!

Reply:
Thanks!